core/source/Time.mint

```diff
@@ -21,185 +36,617 @@ module Time {
     })()
     `
   }
 
-  /* Returns the end of the day / month / week of the given time. */
-  fun endOf (what : String, date : Time) : Time {
+  /*
+  Returns the day of the week of the given time (as a number from 1 to 7).
+
+    Time.dayOfWeekNumber(Time.utcDate(2018, 4, 5)) == 4
+  */
+  fun dayOfWeekNumber (time : Time) : Number {
     `
     (() => {
-      switch (#{what}) {
-        case 'month':
-          return DateFNS.endOfMonth(#{date})
-        case 'week':
-          return DateFNS.endOfWeek(#{date}, { weekStartsOn: 1 })
-        case 'day':
-          return DateFNS.endOfDay(#{date})
-        default:
-          return #{date}
-      }
+      const _ = #{time}.getUTCDay()
+      return _ === 0 ? 7 : _;
     })()
     `
   }
 
-  /* Returns an array of days from the given start to given end time. */
-  fun range (from : Time, to : Time) : Array(Time) {
+  /*
+  Returns the day of week of the given time.
+
+    Time.dayOfWeek(Time.utcDate(2018, 4, 5)) == Weekday::Thursday
+  */
+  fun dayOfWeek (time : Time) : Weekday {
+    case (dayOfWeekNumber(time)) {
+      1 => Weekday::Monday
+      2 => Weekday::Tuesday
+      3 => Weekday::Wednesday
+      4 => Weekday::Thursday
+      5 => Weekday::Friday
+      6 => Weekday::Saturday
+      => Weekday::Sunday
+    }
+  }
+
+  /*
+  Returns the day of month of the given time.
+
+    Time.dayOfMonth(Time.utcDate(2018, 4, 5)) == 5
+  */
+  fun dayOfMonth (time : Time) : Number {
+    `#{time}.getUTCDate()`
+  }
+
+  /*
+  Returns the day of the year of the given time.
+
+    Time.dayOfYear(Time.utcDate(2018, 4, 5)) == 95
+  */
+  fun dayOfYear (time : Time) : Number {
+    `
+    (() => {
+      const first =
+        Date.UTC(#{time}.getUTCFullYear(), 0, 1)
+
+      const current =
+        Date.UTC(#{time}.getUTCFullYear(), #{time}.getUTCMonth(), #{time}.getUTCDate())
+
+      return ((current - first) / 24 / 60 / 60 / 1000) + 1
+    })()
+    `
+  }
+
+  /*
+  Returns the hour of the given time.
+
+    Time.hour(Time.utc(2018, 4, 5, 10, 25, 30, 40) == 10
+  */
+  fun hour (time : Time) : Number {
+    `#{time}.getUTCHours()`
+  }
+
+  /*
+  Returns the minute of the given time.
+
+    Time.minute(Time.utc(2018, 4, 5, 10, 25, 30, 40)) == 25
+  */
+  fun minute (time : Time) : Number {
+    `#{time}.getUTCMinutes()`
+  }
+
+  /*
+  Returns the second of the given time.
+
+    Time.second(Time.utc(2018, 4, 5, 10, 25, 30, 40)) == 30
+  */
+  fun second (time : Time) : Number {
+    `#{time}.getUTCSeconds()`
+  }
+
+  /*
+  Returns the millisecond of the given time.
+
+    Time.millisecond(Time.utc(2018, 4, 5, 10, 25, 30, 40)) == 40
+  */
+  fun millisecond (time : Time) : Number {
+    `#{time}.getUTCMilliseconds()`
+  }
+
+  /*
+  Returns if the year of the given time is a leap year or not.
+
+    Time.isLeapYear(Time.utcDate(2011,1,1)) == false
+    Time.isLeapYear(Time.utcDate(2012,1,1)) == true
+  */
+  fun isLeapYear (time : Time) : Bool {
+    Number.isLeapYear(year(time))
```

Comment:
```suggestion
    time |> year |> Number.isLeapYear
```

Reply:
I think it's more readable my way, this seems to me an overuse of the pipe.

Reply:
I'd assume that any expression with more than 1 call in the chain should be using pipes for expressiveness, but maybe that's just me 🤷‍♂️

Reply:
Is there a performance cost to the pipe, or is it just syntax?
Could we get an arrow that works like Clojure's, without needing a pipe between each step? This would look beautiful as
```
time -> year Time.isLeapYear
```

(Also with regard to the above: why not put `isLeapYear` in the `Time` module? It doesn't seem to have anything to do with Numbers generally.)

Reply:
> Is there a performance cost to the pipe, or is it just syntax?

It's just syntax.

> Could we get an arrow that works like Clojure's, without needing a pipe between each step? This would look beautiful as

I'm working on something that looks like this:

```
time.year().isLeapYear()
```

It's how Nim does it (https://nim-lang.org/docs/manual.html#procedures-method-call-syntax). It allows the use of OO like notation for functions defined in modules which is basically a pipe without the module name.

I'll create a discussion or PR soon for this.